File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jerry/JerryBasicAutonomous.java
package org.firstinspires.ftc.teamcode.jerry;

import com.qualcomm.robotcore.eventloop.opmode.Autonomous;

import org.firstinspires.ftc.teamcode.common.BunyipsOpMode;
import org.firstinspires.ftc.teamcode.common.CameraOp;
import org.firstinspires.ftc.teamcode.common.tasks.TaskImpl;
import org.firstinspires.ftc.teamcode.jerry.config.JerryArm;
import org.firstinspires.ftc.teamcode.jerry.config.JerryConfig;
import org.firstinspires.ftc.teamcode.jerry.config.JerryDrive;
import org.firstinspires.ftc.teamcode.jerry.tasks.JerryBaseDriveTask;

import java.util.ArrayDeque;

@Autonomous(name="<JERRY> POWERPLAY BASIC SIGNAL Autonomous")
public class JerryBasicAutonomous extends BunyipsOpMode {

    private JerryConfig config;
    private CameraOp cam;
    private JerryDrive drive;
    private JerryArm arm;
    private ArrayDeque<TaskImpl> tasks = new ArrayDeque<>();

    @Override
    protected void onInit() {
        config = JerryConfig.newConfig(hardwareMap, telemetry);
        try {
            drive = new JerryDrive(this, config.bl, config.br, config.fl, config.fr);
        } catch (Exception e) {
            telemetry.addLine("Failed to initialise Drive System.");
        }

        tasks.add(new JerryBaseDriveTask(this, 1, drive, 0, 1, 0));

        telemetry.addLine("Ready to go. Parking position has been set to: CENTER");
        telemetry.update();
    }

    @Override
    protected void activeLoop() throws InterruptedException {
        TaskImpl currentTask = tasks.peekFirst();
        if (currentTask == null) {
            return;
        }
        currentTask.run();
        if (currentTask.isFinished()) {
            tasks.removeFirst();
        }
        if (tasks.isEmpty()) {
            drive.deinit();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jerry/JerryBasicAutonomous.kt
package org.firstinspires.ftc.teamcode.jerry

import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import org.firstinspires.ftc.teamcode.common.BunyipsOpMode
import org.firstinspires.ftc.teamcode.common.tasks.TaskImpl
import org.firstinspires.ftc.teamcode.jerry.components.JerryConfig
import org.firstinspires.ftc.teamcode.jerry.components.JerryDrive
import org.firstinspires.ftc.teamcode.jerry.tasks.JerryTimeDriveTask
import java.util.ArrayDeque

@Autonomous(name = "<JERRY> POWERPLAY Drive Forward Autonomous")
class JerryBasicAutonomous : BunyipsOpMode() {
    private var config: JerryConfig? = null
    private var drive: JerryDrive? = null
    private val tasks = ArrayDeque<TaskImpl>()
    override fun onInit() {
        config = JerryConfig.newConfig(hardwareMap, telemetry)
        try {
            drive = JerryDrive(this, config?.bl, config?.br, config?.fl, config?.fr)
        } catch (e: Exception) {
            telemetry.addLine("Failed to initialise Drive System.")
        }
        tasks.add(JerryTimeDriveTask(this, 1.0, drive, 0.0, 1.0, 0.0))
        telemetry.addLine("Ready to go. Parking position has been set to: CENTER")
    }

    @Throws(InterruptedException::class)
    override fun activeLoop() {
        val currentTask = tasks.peekFirst() ?: return
        currentTask.run()
        if (currentTask.isFinished()) {
            tasks.removeFirst()
        }
        if (tasks.isEmpty()) {
            drive?.deinit()
        }
    }
}